### PR TITLE
spann

### PR DESCRIPTION
**Pull request by jobergum**
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: billion-scale-vector-search/README.md [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+* [Docker](https://www.docker.com/) Desktop installed and running. 6GB available memory for Docker is recommended.
+  Refer to [Docker memory](https://docs.vespa.ai/en/operations/docker-containers.html#memory)
+  for details and troubleshooting
+* Operating system: Linux, macOS or Windows 10 Pro (Docker requirement)
+* Architecture: x86_64
+* [Homebrew](https://brew.sh/) to install [Vespa CLI](https://docs.vespa.ai/en/vespa-cli.html), or download
+  a vespa cli release from [Github releases](https://github.com/vespa-engine/vespa/releases).
+* [Java 11](https://openjdk.java.net/projects/jdk/11/) installed.
+* Python3 and numpy to process the vector dataset 
+* [Apache Maven](https://maven.apache.org/install.html) This sample app uses custom Java components and Maven is used
```
Comment (kkraune) [inline billion-scale-vector-search/README.md:29]:
```suggestion
* [Apache Maven](https://maven.apache.org/install.html) - this sample app uses custom Java components and Maven is used
```

File: billion-scale-vector-search/README.md [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+  to build the application. 
+
+Verify Docker Memory Limits:
+
+<pre>
+$ docker info | grep "Total Memory"
+</pre>
+
+Install [Vespa CLI](https://docs.vespa.ai/en/vespa-cli.html).
```
Comment (kkraune) [inline billion-scale-vector-search/README.md:38]:
```suggestion
Install [Vespa CLI](https://docs.vespa.ai/en/vespa-cli.html):
```

File: billion-scale-vector-search/README.md [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+
+<pre >
+$ brew install vespa-cli
+</pre>
+
+Set target env, it's also possible to deploy this app to [Vespa Cloud](https://cloud.vespa.ai/)
+using target cloud. For Vespa cloud deployments to [perf env](https://cloud.vespa.ai/en/reference/zones.html) 
+replace the [src/main/application/services.xml](src/main/application/services.xml) with 
+[src/main/application/services.xml](src/main/application/cloud-services.xml).
```
Comment (kkraune) [inline billion-scale-vector-search/README.md:47]:
```suggestion
[src/main/application/cloud-services.xml](src/main/application/cloud-services.xml). 
```

File: billion-scale-vector-search/README.md [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+<pre >
+$ brew install vespa-cli
+</pre>
+
+Set target env, it's also possible to deploy this app to [Vespa Cloud](https://cloud.vespa.ai/)
+using target cloud. For Vespa cloud deployments to [perf env](https://cloud.vespa.ai/en/reference/zones.html) 
+replace the [src/main/application/services.xml](src/main/application/services.xml) with 
+[src/main/application/services.xml](src/main/application/cloud-services.xml). 
+
+For local deployment using docker image use
```
Comment (kkraune) [inline billion-scale-vector-search/README.md:49]:
```suggestion
For local deployment using docker image use:
```

File: billion-scale-vector-search/README.md [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+
+<pre data-test="exec">
+$ vespa config set target local
+</pre>
+
+For cloud deployment using [Vespa Cloud](https://cloud.vespa.ai/) use
```
Comment (kkraune) [inline billion-scale-vector-search/README.md:55]:
```suggestion
For cloud deployment using [Vespa Cloud](https://cloud.vespa.ai/) use:
```

File: billion-scale-vector-search/README.md [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+  vespaengine/vespa
+</pre>
+
+Verify that configuration service (deploy api) is ready
+
+<pre data-test="exec">
+$ vespa status deploy --wait 300
+</pre>
+
+Download this sample application
```
Comment (kkraune) [inline billion-scale-vector-search/README.md:85]:
```suggestion
Download this sample application:
```

File: billion-scale-vector-search/README.md [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+
+
+## Download Vector Data
+
+This sample app uses the Microsoft SPACEV vector dataset from 
+[https://big-ann-benchmarks.com/](https://big-ann-benchmarks.com/) 
+
+This quick start uses the first 10M vectors of the 100M slice sample. 
+
+This sample file is about 1GB (10M vectors)
```
Comment (kkraune) [inline billion-scale-vector-search/README.md:99]:
```suggestion
It uses the first 10M vectors of the 100M slice sample.
This sample file is about 1GB (10M vectors):
```

File: billion-scale-vector-search/README.md [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+  https://data.vespa.oath.cloud/sample-apps-data/spacev10m_base.i8bin
+</pre>
+
+Generate the feed file for the first 10M vectors from the 100M sample. 
+This step creates two feed files:
+
+* `graph-vectors.jsonl`
+* `if-vectors.jsonl`
+
+Install python dependencies
```
Comment (kkraune) [inline billion-scale-vector-search/README.md:112]:
```suggestion
Install python dependencies:
```

File: billion-scale-vector-search/README.md [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+$ pip3 install numpy requests tqdm 
+</pre>
+
+<pre data-test="exec">
+$ python3 src/main/python/create-vespa-feed.py spacev10m_base.i8bin
+</pre>
+
+## Build and deploy Vespa app 
+
+Build the sample app
```
Comment (kkraune) [inline billion-scale-vector-search/README.md:124]:
```suggestion
Build the sample app:
```

File: billion-scale-vector-search/README.md [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+$ vespa deploy --wait 300
+</pre>
+
+Wait for the application endpoint to become available
+
+<pre data-test="exec">
+$ vespa status --wait 300
+</pre>
+
+Test basic functionality
```
Comment (kkraune) [inline billion-scale-vector-search/README.md:142]:
```suggestion
Test basic functionality:
```

File: billion-scale-vector-search/README.md [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+
+Feed the vector using the [vespa-feed-client](https://docs.vespa.ai/en/vespa-feed-client.html):
+
+<pre data-test="exec">
+$ curl -L -o vespa-feed-client-cli.zip \
+    https://search.maven.org/remotecontent?filepath=com/yahoo/vespa/vespa-feed-client-cli/7.588.57/vespa-feed-client-cli-7.588.57-zip.zip
+$ unzip -o vespa-feed-client-cli.zip
+</pre>
+
+The graph vectors must be feed before the if vectors:
```
Comment (kkraune) [inline billion-scale-vector-search/README.md:157]:
```suggestion
The _graph_ vectors must be feed before the _if_ vectors:
```

File: billion-scale-vector-search/README.md [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+
+<pre data-test="exec">
+$ ./vespa-feed-client-cli/vespa-feed-client \
+  --file graph-vectors.jsonl --endpoint http://localhost:8080/
+$ ./vespa-feed-client-cli/vespa-feed-client \
+  --file if-vectors.jsonl --endpoint http://localhost:8080/
+</pre>
+
+## Recall Evaluation
+Download the query vectors and the ground truth for the 10M first vectors
```
Comment (kkraune) [inline billion-scale-vector-search/README.md:167]:
```suggestion
Download the query vectors and the ground truth for the 10M first vectors:
```

File: billion-scale-vector-search/README.md [hunk 13]
```diff
@@ -0,0 +1,196 @@
...
+## Recall Evaluation
+Download the query vectors and the ground truth for the 10M first vectors 
+<pre data-test="exec">
+$ curl -L -o query.i8bin \
+https://comp21storage.blob.core.windows.net/publiccontainer/comp21/spacev1b/query.i8bin
+$ curl -L -o spacev10m_gt100.i8bin \
+  https://data.vespa.oath.cloud/sample-apps-data/spacev10m_gt100.i8bin
+</pre>
+
+Run first 1K queries and evaluate recall@10. Higher number of clusters gives higher recall
```
Comment (kkraune) [inline billion-scale-vector-search/README.md:175]:
```suggestion
Run first 1K queries and evaluate recall@10. Higher number of clusters gives higher recall:
```

File: billion-scale-vector-search/pom.xml [hunk 14]
```diff
@@ -0,0 +1,22 @@
+<?xml version="1.0"?>
+<!-- Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root. -->
```
Comment (kkraune) [inline billion-scale-vector-search/pom.xml:2]:
```suggestion
<!-- Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root. -->
```

**Review by bratseth — APPROVED** [review]
LGTM.

It's not necessary for ClusteringComponent to be a component any more since it doesn't have any state. If you want I can do a pass and make it a bit more idiomatic and add some tests.

**Comment by jobergum** [comment]
That would be fantastic @bratseth.

**Pull request merged**